supplier_server_mock/supplier_server/app.py

```diff
@@ -14,156 +14,201 @@
-    utils.check_non_timeslot_product(product_id)
     start = utils.get_date(request.args, 'start')
     end = utils.get_date(request.args, 'end')
     if start > end:
         raise exceptions.BadRequest(2001, 'Incorrect date range', 'The end date cannot be earlier than start date')
     today = date.today()
     if end < today:
-        return jsonify([])
+        return jsonify({})
     start = max(today, start)
-    result = []
+    result = {}
     day = start
     while day <= end:
-        result.append(utils.get_availability(product_id, day))
+        result[str(day)] = utils.get_availability(product_id, day)
         day = day + timedelta(days=1)
     return jsonify(result)
 
 
-@app.route('/v1/products/<product_id>/timeslots')
+@app.route('/v2/products/<product_id>/reservation', methods=['POST'])
 @authorization_header
-@date_range_validator
-def availability_timeslots(product_id: str):
+def reservation(product_id: str):
     utils.check_product_id(product_id)
-    utils.check_timeslot_product(product_id)
-    start = utils.get_date(request.args, 'start')
-    end = utils.get_date(request.args, 'end')
-    if start > end:
-        raise exceptions.BadRequest(2001, 'Incorrect date range', 'The end date cannot be earlier than start date')
-    today = date.today()
-    if end < today:
-        return jsonify([])
-    start = max(today, start)
-    results = []
-    day = start
-    while day <= end:
-        for timeslot in ('17:30', '19:30'):
-            result = utils.get_availability(product_id, day)
-            result['start'] = timeslot
-            result['end'] = arrow.get(f'2000-01-01 {timeslot}').shift(hours=1).strftime('%H:%M')
-            results.append(result)
-        day = day + timedelta(days=1)
-    return jsonify(results)
 
+    datetime_parameter_value = request.json.get('datetime')
+    if not datetime_parameter_value:
+        raise exceptions.BadRequest(1000, 'Missing argument', 'Required argument "datetime" was not found')
+
+    tickets = request.json.get('tickets')
+    if not tickets:
+        raise exceptions.BadRequest(1000, 'Missing argument', 'Required argument "tickets" was not found')
+
+    customer = request.json.get('customer')
+    if not customer:
+        raise exceptions.BadRequest(1000, 'Missing argument', 'Required argument "customer" was not found')
+
+    # get the additional order-level information required by this product
+    product_requires_additional_order_data = utils.product_required_additional_order_data(product_id)
+    required_order_data = request.json.get('required_order_data')
+    if product_requires_additional_order_data:
+        if not required_order_data or set(product_requires_additional_order_data) != set(required_order_data.keys()):
+            raise exceptions.BadRequest(
+                1003,
+                'Missing required fields',
+                f'Missing required additional order data: {",".join(product_requires_additional_order_data)}',
```

Comment:
Small improvements:
`utils.product_required_additional_order_data(product_id)` could already return a set

and then

```py
required_order_data_keys = required_order_data.keys()
missing_fields = product_requires_additional_order_data.difference(required_order_data_keys)
if not required_order_data or missing_fields:
    raise exceptions.BadRequest(
        1003,
        'Missing required fields',
        f'Missing required additional order data: {sorted(missing_fields)}',
    )
```